### ISSUE TITLE
pysqlite3 doesn't support asyncio

### ISSUE THREAD

**Issue by nickleman** [talkpython/web-applications-with-fastapi-course#4]
It seems the latest sqlalchemy beta release now does as better check for sqlite async support (or you have a different version of sqlite). I'm using python 3.8 and when I get to Chapter 8 I get the following error message after adding create_async_engine, etc.

sqlalchemy.exc.InvalidRequestError: The asyncio extension requires an async driver to be used. The loaded 'pysqlite' is not async.

**Comment by mikeckennedy** [comment]
Thanks for the heads up. I'm glad I pinned the version there to one that worked. Do you know of a fix? From what I see they seem to say it's just not available yet.

https://stackoverflow.com/questions/66431083/using-asyncio-extension-with-sqlite-backend-broken-by-version-upgrade

**Comment by nickleman** [comment]
That's what I'm seeing too. There is an async sqlite3 library, but it does not appear to be the one that sqlalchemy/asyncio uses. It's called aiosqlite and is already used by the "databases" module I believe, but sqlalchemy does not seem to use that, they are using the standard sqlite3 which is not async ready yet. From what I read, the app you built would not actually be async since the sqlite library isn't, but you'd only notice that under load. :) But as a proof of concept and practice app, it works fine. I now need to back out to the earlier version, although, I thought I actually installed the same version you did, I wonder if they pushed an update under the same version number.

**Comment by nickleman** [comment]
I take it back I have 1.4.0b3 installed. Which, BTW, wouldn't work with the version number setup you created in the sample app which only accepts 3 integers. I've always had to handle them as strings and use packaging.version when I needed to do comparisons between them... Version strings can be very obnoxious to deal with unfortunately.

**Comment by nickleman** [comment]
FYI... 1.4.0b2 still works.

**Comment by mikeckennedy** [comment]
Thanks @nickleman I spoke with the SQLAlchemy folks and they are working on the SQLite driver to support the asyncio layer eventually. For now, just pin the version and I'll let you know here when it's fully supported. 

The goal of the async / await section with SQLite wasn't really to get async behavior. But if you just toggle the connection string to say Postgres in a real production situation, it should come alive. We're just trying to avoid having you all setup and configure a separate server for the coruse.

**Comment by ruimartinsptl** [comment]
I'm not doing this course, but I had the same problem, and googling it I saw this thread.

Then, I solve my problem replacing:

```python
engine = create_async_engine(
        "sqlite+pysqlite:///:memory:",
        echo=True,
    )
```

with
```python
engine = create_async_engine(
        "sqlite+aiosqlite:///:memory:",
        echo=True,
    )
```

requirements.txt:
```text
SQLAlchemy==1.4.3
aiosqlite==0.17.0
```

Python 3.9.2

---

I hope this commentary is useful for you.

**Comment by mikeckennedy** [comment]
Thank you all. @ruimartinsptl I appreciate your pointing out this works now. I had been talking with Mike Bayer (creator of SQLAlchemy) and he gave me the heads up that this fix was coming but we were waiting on aiosqlite to update. Apparently it released now.

**Comment by SevaShpun** [comment]
@ruimartinsptl 
Thanks! It worked!

**Comment by mikeckennedy** [comment]
Hi all. Yes, unfortunately when we recorded this, it worked fine. To "help us" SQLAlchemy made this an arbitrary breaking change. I did update the code in this repo when that happened and put a descriptive comment on why:

https://github.com/talkpython/web-applications-with-fastapi-course/blob/f197a394085547ed4a4367bc617de5c6e3c3830d/code/ch8-async-databases/data/db_session.py#L28

@ruimartinsptl Thanks for helping out.
